File: mockwebserver/src/main/java/com/squareup/okhttp/mockwebserver/rule/MockWebServerRule.java
/*
 * Copyright (C) 2014 Square, Inc.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *      http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
package com.squareup.okhttp.mockwebserver.rule;

import com.squareup.okhttp.HttpUrl;
import com.squareup.okhttp.mockwebserver.MockResponse;
import com.squareup.okhttp.mockwebserver.MockWebServer;
import com.squareup.okhttp.mockwebserver.RecordedRequest;
import java.io.IOException;
import java.net.URL;
import java.util.logging.Level;
import java.util.logging.Logger;
import org.junit.rules.ExternalResource;

/**
 * Allows you to use {@link MockWebServer} as a JUnit test rule.
 *
 * <p>This rule starts {@link MockWebServer} on an available port before your test runs, and shuts
 * it down after it completes.
 */
public class MockWebServerRule extends ExternalResource {
  private static final Logger logger = Logger.getLogger(MockWebServerRule.class.getName());

  private final MockWebServer server = new MockWebServer();
  private boolean started;

  @Override protected void before() {
    if (started) return;
    started = true;
    try {
      server.start();
    } catch (IOException e) {
      throw new RuntimeException(e);
    }
  }

  @Override protected void after() {
    try {
      server.shutdown();
    } catch (IOException e) {
      logger.log(Level.WARNING, "MockWebServer shutdown failed", e);
    }
  }

  public String getHostName() {
    if (!started) before();
    return server.getHostName();
  }

  public int getPort() {
    if (!started) before();
    return server.getPort();
  }

  public int getRequestCount() {
    return server.getRequestCount();
  }

  public void enqueue(MockResponse response) {
    server.enqueue(response);
  }

  public RecordedRequest takeRequest() throws InterruptedException {
    return server.takeRequest();
  }

  @Deprecated
  public URL getUrl(String path) {
    return server.getUrl(path);
  }

  public HttpUrl url(String path) {
    return server.url(path);
  }

  /** For any other functionality, use the {@linkplain MockWebServer} directly. */
  public MockWebServer get() {
    return server;
  }
}

File: mockwebserver-junit4/src/main/kotlin/mockwebserver3/junit4/MockWebServerRule.kt
/*
 * Copyright (C) 2020 Square, Inc.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *      http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
package mockwebserver3.junit4

import mockwebserver3.MockWebServer
import org.junit.rules.ExternalResource
import java.io.IOException
import java.util.logging.Level
import java.util.logging.Logger

/**
 * Runs MockWebServer for the duration of a single test method.
 *
 * In Java JUnit 4 tests (ie. tests annotated `@org.junit.Test`), use this by defining a field with
 * the `@Rule` annotation:
 *
 * ```
 * @Rule public final MockWebServerRule serverRule = new MockWebServerRule();
 * ```
 *
 * For Kotlin the `@JvmField` annotation is also necessary:
 *
 * ```
 * @JvmField @Rule val serverRule = MockWebServerRule()
 * ```
 */
class MockWebServerRule : ExternalResource() {
  val server: MockWebServer = MockWebServer()

  override fun before() {
    try {
      server.start()
    } catch (e: IOException) {
      throw RuntimeException(e)
    }
  }

  override fun after() {
    try {
      server.shutdown()
    } catch (e: IOException) {
      logger.log(Level.WARNING, "MockWebServer shutdown failed", e)
    }
  }

  companion object {
    private val logger = Logger.getLogger(MockWebServerRule::class.java.name)
  }
}
